qiime2/metadata/metadata.py

```diff
@@ -362,6 +386,43 @@ def __init__(self, dataframe):
         super().__init__(dataframe.index)
 
         self._dataframe, self._columns = self._normalize_dataframe(dataframe)
+        self.contains_renamed_columns = False
+
+    @property
+    def hash(self):
+        if not isinstance(self._hash, uuid.UUID):
+            return f'{self._hash:x}'
+        else:
+            return self._hash
+
+    @property
+    def _hash(self):
+        hash = None
+        if self.artifacts:
```

Comment:
One (subjective) pattern that I think improves readability is to go from this:

```python
if self.artifacts:
    hash = self.artifacts[0].uuid
    for artifact in self.artifacts[1:]:
        hash = int(hash) ^ int(artifact.uuid)
```

to this:

```python
if self.artifacts:
    artifacts = iter(self.artifacts)

    # always at least one here
    hash = getattr(next(artifacts), 'uuid')

    for artifact in artifacts:
        hash = int(hash) ^ int(artifact.uuid)
```

YMMV.